### PR TITLE
store: add HBase support

### PR DESCRIPTION
**Pull request by disksing**
This PR aims to make HBase as the underlying transactional KV engine for TiDB.
We use a modified version of XiaoMi/Themis (https://github.com/pingcap/themis) to support cross-row transaction on HBase.

Notice: HBase 0.98.4+ only.

### CODE REVIEW

File: store/hbase/snapshot.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+func newHbaseSnapshot(txn *themis.Txn, storeName string) *hbaseSnapshot {
+	return &hbaseSnapshot{
+		txn:       txn,
+		storeName: storeName,
+		cache:     make(map[string][]byte),
+	}
+}
+
+func (s *hbaseSnapshot) Get(k kv.Key) ([]byte, error) {
+	if b, ok := s.cache[string(k)]; ok {
```
Comment (c4pt0r) [inline store/hbase/snapshot.go]: Please add some comments.

File: Makefile [hunk 2]
```diff
@@ -19,6 +19,8 @@ godep:
 	go get github.com/tools/godep
 
 build:
+	go get github.com/pingcap/go-hbase
```
Comment (qiuyesuifeng) [inline Makefile]: How about moving `go get ...` to `godep` section above?

File: metric/metric.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	r    = metrics.NewRegistry()
+	once sync.Once
+)
+
+// Register registers a new metric for observation.
+func Register(name string, m interface{}) {
+	r.Register(name, m)
+}
+
+// Inc increments specific counter metric.
```
Comment (qiuyesuifeng) [inline metric/metric.go]: increases

File: metric/metric_test.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	testMetricName     = "test-metric"
+	testTimeMetricName = "time-metric"
+)
+
+func (t *testSuite) TestRegMetric(c *C) {
+	Register(testMetricName, metrics.NewCounter())
+	Register(testMetricName, metrics.NewHistogram(metrics.NewUniformSample(1000)))
+
+	v := r.Get(testMetricName)
+	c.Assert(v != nil, Equals, true)
```
Comment (qiuyesuifeng) [inline metric/metric_test.go]: c.Assert(v, NotNil)

File: store/hbase/kv.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hbasekv
+
+import (
+	"strings"
+
```
Comment (qiuyesuifeng) [inline store/hbase/kv.go]: Remove empty line.

File: store/hbase/kv.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+	if len(zkInfo) == 0 {
+		return nil, ErrZkInfoErr
+	}
+	if store, ok := mc.cache[zkInfo]; ok {
+		// TODO: check the cache store has the same engine with this Driver.
+		return store, nil
+	}
+	zks := strings.Split(zkInfo, ",")
+	if len(zks) == 0 {
```
Comment (qiuyesuifeng) [inline store/hbase/kv.go]: len(zks) will not be 0, so seems this check is useless?

File: store/hbase/kv.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+const (
+	// ColFamily is the hbase column family name.
+	ColFamily = "f"
+	// Qualifier is the hbase column name.
+	Qualifier = "q"
+	// FmlAndQual is a shortcut.
+	FmlAndQual = ColFamily + ":" + Qualifier
+	// HbaseTableName is the table name in hbase which is invisible to SQL layer.
+	HbaseTableName = "tidb"
```
Comment (qiuyesuifeng) [inline store/hbase/kv.go]: Is this variable only used in this package, if so, it can be unexported.

File: store/hbase/snapshot.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+const hbaseBatchSize = 1000
+
+// hbaseSnapshot implements MvccSnapshot interface.
+type hbaseSnapshot struct {
+	txn       *themis.Txn
+	storeName string
+	cache     map[string][]byte
+}
+
+// newHBaseSnapshot creates a snapshot of an HBase store.
```
Comment (qiuyesuifeng) [inline store/hbase/snapshot.go]: a HBase store

File: store/hbase/snapshot.go [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+// Get gets the value for key k from snapshot. It internally uses a cache to
+// reduce RPC calls. If cache not hit, it uses a Scanner to query a batch of
+// data.
+func (s *hbaseSnapshot) Get(k kv.Key) ([]byte, error) {
+	if b, ok := s.cache[string(k)]; ok {
+		return b, nil
+	}
+
+	s.cache = make(map[string][]byte)
```
Comment (qiuyesuifeng) [inline store/hbase/snapshot.go]: A little confused, why make a new map?

File: store/hbase/snapshot.go [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+
+type hbaseIter struct {
+	*themis.ThemisScanner
+	rs    *hbase.ResultRow
+	valid bool
+}
+
+func (it *hbaseIter) Next() error {
+	it.rs = it.ThemisScanner.Next()
+	it.valid = it.rs != nil && len(it.rs.Columns) > 0 && !it.ThemisScanner.Closed()
```
Comment (qiuyesuifeng) [inline store/hbase/snapshot.go]: What is this line code? Seems strange.

File: util/batch_worker.go [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+	var jobs []interface{}
+	b.mu.Lock()
+	b.pendingJobs.PushBack(job)
+	if b.pendingJobs.Len() >= b.batchSize {
+		jobs = make([]interface{}, 0, b.batchSize)
+		// pop first batchSize jobs to workerFn
+		for i := 0; i < b.batchSize; i++ {
+			ele := b.pendingJobs.Front()
+			jobs = append(jobs, ele.Value)
+			b.pendingJobs.Remove(ele)
```
Comment (qiuyesuifeng) [inline util/batch_worker.go]: `Remove removes e from l if e is an element of list l. It returns the element value e.Value.` 
So can call remove first?

Comment (qiuyesuifeng) [inline util/batch_worker.go]: Ignore it, seems not.

File: util/batch_worker.go [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+	return &BatchWorker{
+		pendingJobs: list.New(),
+		batchSize:   batchSize,
+		fn:          fn,
+	}
+}
+
+// Submit submits a job to BatchWorker.
+func (b *BatchWorker) Submit(job interface{}) {
+	var jobs []interface{}
```
Comment (qiuyesuifeng) [inline util/batch_worker.go]: I prefer to define `jobs := make([]interface{}, 0, b.batchSize)` directly.

File: store/hbase/txn.go [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+func (txn *hbaseTxn) String() string {
+	return fmt.Sprintf("%d", txn.tid)
+}
+
+func (txn *hbaseTxn) Get(k kv.Key) ([]byte, error) {
+	log.Debugf("get key:%q, txn:%d", k, txn.tid)
+	k = kv.EncodeKey(k)
+	val, err := txn.UnionStore.Get(k)
+	if kv.IsErrNotFound(err) || len(val) == 0 {
+		return nil, kv.ErrNotExist
```
Comment (qiuyesuifeng) [inline store/hbase/txn.go]: It is better to use `errors.Trace(kv.ErrNotExist)`, for many places return `kv.ErrNotExist`, so we should use juju/errors to trace error stack info.

File: store/hbase/txn.go [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+	"github.com/ngaut/log"
+	"github.com/pingcap/go-hbase"
+	"github.com/pingcap/go-themis"
+	"github.com/pingcap/tidb/kv"
+)
+
+var (
+	_ kv.Transaction = (*hbaseTxn)(nil)
+	// ErrInvalidTxn is the error when commits or rollbacks in an invalid transaction.
+	ErrInvalidTxn = errors.New("invalid transaction")
```
Comment (qiuyesuifeng) [inline store/hbase/txn.go]: Seems this variable has already been defined in localstore package, can be merged later?

File: store/hbase/txn.go [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+		return ErrInvalidTxn
+	}
+	log.Warnf("Rollback txn %d", txn.tid)
+	return txn.close()
+}
+
+func (txn *hbaseTxn) LockKeys(keys ...kv.Key) error {
+	for _, key := range keys {
+		key = kv.EncodeKey(key)
+		if err := txn.Txn.LockRow(txn.storeName, key); err != nil {
```
Comment (qiuyesuifeng) [inline store/hbase/txn.go]: What is the meaning of LockRow?

File: store/hbase/txn.go [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+
+func (txn *hbaseTxn) doCommit() error {
+	err := txn.each(func(iter kv.Iterator) error {
+		var row, val []byte
+		row = make([]byte, len(iter.Key()))
+		if len(iter.Value()) == 0 { // Deleted marker
+			copy(row, iter.Key())
+			d := hbase.NewDelete(row)
+			d.AddStringColumn(ColFamily, Qualifier)
+			txn.Txn.Delete(txn.storeName, d)
```
Comment (qiuyesuifeng) [inline store/hbase/txn.go]: Any need to process error?

File: Makefile (hunk 2 above)
Comment (disksing) [inline Makefile]: Got it. I was thinking `godep` means the tool gedep literally...

File: store/hbase/snapshot.go [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+	return m, nil
+}
+
+// RangeGet implements kv.Snapshot.RangeGet().
+// The range should be [start, end] as Snapshot.RangeGet() indicated.
+func (s *hbaseSnapshot) RangeGet(start, end kv.Key, limit int) (map[string][]byte, error) {
+	scanner := s.txn.GetScanner([]byte(s.storeName), start, end, limit)
+	defer scanner.Close()
+
+	m := make(map[string][]byte)
```
Comment (ngaut) [inline store/hbase/snapshot.go]: Make map with capability limit.

File: kv/cache_snapshot.go [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+	}
+}
+
+// Get gets value from snapshot and saves it in cache.
+func (c *cacheSnapshot) Get(k Key) ([]byte, error) {
+	v, err := c.cache.Get(k)
+	if IsErrNotFound(err) {
+		v, err = c.snapshot.Get(k)
+		if err == nil {
+			c.cache.Set([]byte(k), v)
```
Comment (qiuyesuifeng) [inline kv/cache_snapshot.go]: Any need to process return error？

File: kv/cache_snapshot.go [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return v, nil
+}
+
+// BatchGet gets a batch of values from snapshot and saves them in cache.
+func (c *cacheSnapshot) BatchGet(keys []Key) (map[string][]byte, error) {
+	m := make(map[string][]byte)
+	var missKeys []Key
```
Comment (qiuyesuifeng) [inline kv/cache_snapshot.go]: I prefer `missedKeys := []Key{}`

Comment (ngaut) [inline kv/cache_snapshot.go]: No, that's bad style.

File: stmt/stmts/insert_test.go [hunk 20]
```diff
@@ -96,3 +96,30 @@ func (s *testStmtSuite) TestInsert(c *C) {
 	_, err = s.testDB.Exec(`insert into insert_test (id, c2) values(1, 1) on duplicate key update t.c2 = 10`)
 	c.Assert(err, NotNil)
 }
+
+func (s *testStmtSuite) TestInsertAutoInc(c *C) {
+	createSQL := `drop table if exists insert_autoinc_test; create table insert_autoinc_test (id int auto_increment, c1 int);`
```
Comment (qiuyesuifeng) [inline stmt/stmts/insert_test.go]: Seems MySQL need primary key to set auto_increment spec?
This create table statement should be failed.

File: store/hbase/snapshot.go [hunk 21]
```diff
@@ -0,0 +1,199 @@
...
+	m := make(map[string][]byte, len(rows))
+	for _, r := range rows {
+		k := string(r.Row)
+		v := r.Columns[hbaseFmlAndQual].Value
+		m[k] = v
+	}
+	return m, nil
+}
+
+// RangeGet implements kv.Snapshot.RangeGet().
```
Comment (qiuyesuifeng) [inline store/hbase/snapshot.go]: Comment should be `RangeGet implements kv.Snapshot.RangeGet interface.`

File: store/localstore/snapshot.go [hunk 22]
```diff
@@ -108,6 +108,41 @@ func (s *dbSnapshot) Get(k kv.Key) ([]byte, error) {
 	return s.MvccGet(k, s.version)
 }
 
+func (s *dbSnapshot) BatchGet(keys []kv.Key) (map[string][]byte, error) {
+	m := make(map[string][]byte)
+	for _, k := range keys {
+		v, err := s.Get(k)
+		if err != nil && !kv.IsErrNotFound(err) {
+			return nil, err
```
Comment (qiuyesuifeng) [inline store/localstore/snapshot.go]: errors.Trace(err)

File: store/hbase/snapshot.go (hunk 17 above)
Comment (disksing) [inline store/hbase/snapshot.go]: I suppose if the number of keys between start and end is significantly smaller than limit, making a big map maybe a huge waste.

File: stmt/stmts/insert_test.go (hunk 20 above)
Comment (disksing) [inline stmt/stmts/insert_test.go]: #586

Comment (qiuyesuifeng) [inline stmt/stmts/insert_test.go]: Gotcha.

**Comment by ngaut** [comment]
LGTM

**Comment by c4pt0r** [comment]
@siddontang  PTAL

**Comment by siddontang** [comment]
LGTM

**Pull request merged**